xls/modules/rle/rle_dec.x

```diff
@@ -0,0 +1,274 @@
+// (output channel)     ──────────────────┘     └─────────────────┘     └────
+
+
+import std
+
+pub struct DecInData<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    symbol: bits[SYMBOL_WIDTH], // compressed symbol
+    count: bits[COUNT_WIDTH],   // symbol count
+    last: bool,                 // if this is the last symbol
+}
+
+pub struct DecOutData<SYMBOL_WIDTH: u32> {
+    symbol: bits[SYMBOL_WIDTH], // input symbol to compress
+    last: bool,                 // if this is the last symbol
+}
+
+// structure to preserve the state of an RLE decoder
+struct RunLengthDecoderState<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+  // symbol to be repeated on output
+  symbol: bits[SYMBOL_WIDTH],
+  // count of symbols that has to be send
+  count: bits[COUNT_WIDTH],
+  // send last when repeat ends
+  last: bool,
+}
+ // RLE decoder implementation
+pub proc RunLengthDecoder<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+  input_r: chan<DecInData<SYMBOL_WIDTH, COUNT_WIDTH>> in;
+  output_s: chan<DecOutData<SYMBOL_WIDTH>> out;
+
+  init {(
+    RunLengthDecoderState<SYMBOL_WIDTH, COUNT_WIDTH> {
+      symbol: bits[SYMBOL_WIDTH]:0,
+      count:  bits[COUNT_WIDTH]:0,
+      last:   bool:false,
+    }
+  )}
+
+  config (
+    input_r: chan<DecInData<SYMBOL_WIDTH, COUNT_WIDTH>> in,
+    output_s: chan<DecOutData<SYMBOL_WIDTH>> out,
+  ) {(input_r, output_s)}
+
+  next (tok: token, state: RunLengthDecoderState<SYMBOL_WIDTH, COUNT_WIDTH>) {
+    let state_input = DecInData {
+      symbol: state.symbol,
+      count: state.count,
+      last: state.last
+    };
+    let recv_next_symbol = (state.count == bits[COUNT_WIDTH]:0);
+    let (tok, input) = recv_if(tok, input_r, recv_next_symbol, state_input);
+    let next_count = input.count - bits[COUNT_WIDTH]:1;
+    let done_sending = (next_count == bits[COUNT_WIDTH]:0);
+    let send_last = input.last && done_sending;
+    let data_tok = send(tok, output_s, DecOutData {
+      symbol: input.symbol,
+      last: send_last
+    });
+    if (send_last) {
+      zero!<RunLengthDecoderState>()
+    } else {
+      RunLengthDecoderState {
+        symbol: input.symbol,
+        count: next_count,
+        last: input.last,
+      }
+    }
+  }
+}
+
+
+// RLE decoder specialization for the codegen
+proc RunLengthDecoder32 {
+  init {()}
+
+  config (
+    input_r: chan<DecInData<32, 2>> in,
+    output_s: chan<DecOutData<32>> out,
+  ) {
+    spawn RunLengthDecoder<u32:32, u32:2>(input_r, output_s);
+    ()
+  }
+
+  next (tok: token, state: ()) {
+    ()
+  }
+}
+
+// Tests
+
+const TEST_SYMBOL_WIDTH = u32:32;
+const TEST_COUNT_WIDTH  = u32:32;
+
+type TestSymbol     = bits[TEST_SYMBOL_WIDTH];
+type TestCount      = bits[TEST_COUNT_WIDTH];
+type TestStimulus   = (TestSymbol, TestCount);
+type TestDecInData  = DecInData<TEST_SYMBOL_WIDTH, TEST_COUNT_WIDTH>;
+type TestDecOutData = DecOutData<TEST_SYMBOL_WIDTH>;
+
+// Check that RLE decoder will create 2 `last` output packets,
```

Comment:
q: are there other corner case we should test for?

Reply:
I think there are no more corner cases to be checked.

Reply:
what if send a packet with a `count` of `0`?

Reply:
I assumed that count is never 0, as encoder never emits pairs with count = 0.

Reply:
I see, that makes sense if we only care about testing valid input.

Reply:
I think it is useful to treat count as being offset by 1 so count=0 means 1 symbol, count=1 means 2 symbols, etc. Disallowing count=0 is wasteful when count is narrow, e.g. in the case of count being 1 bit it will always be longer. If you encode count with an offset, it should probably be named something like `offset_count` or `count_minus_one` to avoid confusion.

Reply:
Regardless, as-is I think there should be a `fail!()` if count is 0, as otherwise it is interpreted as `2^COUNT_WIDTH` symbols.

Reply:
If you have nothing against it, for now I will add `fail!()` if count is `0`.

Reply:
@mtdudek do you think we should file an issue to capture @grebe's offset by 1 suggestion? I think it would be worth discussing it.

Reply:
filed https://github.com/google/xls/issues/1070